common/src/main/java/com/github/khanshoaib3/minecraft_access/features/inventory_controls/GroupGenerator.java

```diff
@@ -467,8 +466,51 @@ else if (screen.getHandler() instanceof HopperScreenHandler)
             foundGroups.add(itemOutputGroup);
         }
 
-        if (unknownGroup.slotItems.size() > 0) {
-            foundGroups.add(unknownGroup);
+        // Adjacency is used here instead of slot.inventory because some mods have non-adjacent slots in the same
+        // inventory and adjacent slots in different inventories
+        List<List<SlotItem>> unknownGroups = new ArrayList<>(unknownSlots.size());
+        for (SlotItem slot : unknownSlots) {
+            unknownGroups.stream()
+                    .filter(group ->
+                            group.stream()
+                                    .anyMatch(groupSlot ->
+                                            groupSlot.x == slot.x && (groupSlot.y == slot.y + 18 || groupSlot.y == slot.y - 18)
+                                            || groupSlot.y == slot.y && (groupSlot.x == slot.x + 18 || groupSlot.x == slot.x - 18)
+                                    )
+                    )
+                    .findFirst()
+                    .or(() -> {
+                        List<SlotItem> group = new ArrayList<>(unknownSlots.size());
+                        unknownGroups.add(group);
+                        return Optional.of(group);
+                    })
+                    .get()
+                    .add(slot);
+        }
+        Map<String, Byte> usedNames = new HashMap<>(unknownGroups.size());
+        for (List<SlotItem> group : unknownGroups) {
+            String groupName;
+            if (group.stream()
+                    .map(slot -> slot.slot.getClass().getSimpleName())
+                    .distinct()
+                    .limit(2)
+                    .count() == 1) {
+                groupName = CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, group.get(0).slot.getClass().getSimpleName());
+		// Don't use vanilla obfuscated class names
+                if (groupName.startsWith("class_")) {
+                    groupName = "unknown_group";
+                }
+            } else {
+                groupName = "unknown_group";
+            }
+            if (usedNames.containsKey(groupName)) {
+                byte n = (byte) (usedNames.get(groupName) + 1);
+                usedNames.put(groupName, n);
+                groupName += String.format("_%d", n);
```

Comment:
Really nice idea about getting slot group names.
It misses concerns about I18N but I can't come up with better idea to amend it.

Reply:
Some kind of I18N would be a good idea though so that it doesn't say `minecraft_access.slot_group.<name> group` even if it's not proper I18N and it just says the (usually) English name, but I dont have enough experience with I18N to implement such a thing.